### ISSUE TITLE
--set-samba-secret option not useable with samba 4.7 ?

### ISSUE THREAD

**Issue by jaroslawp** [msktutil/msktutil#124]
The --set-samba-secret option seems to be not useable with samba 4.7 (possibly 4.6 too ?..).

The implementation of "net changesecretpw" seem to have changed: on older samba versions (4.2) this command was just storing the password in samba's secrets.tdb, but the same command in samba 4.7 performs additional checks .. which fail on a system which has not 'joined' the Active Directory:

# net changesecretpw -f
Enter machine password: 
secrets_fetch_or_upgrade_domain_info: secrets_fetch_domain_sid(DOMAIN) failed
secrets_prepare_password_change: secrets_fetch_or_upgrade_domain_info(DOMAIN) failed
Unable to write the machine account password in the secrets database

**Comment by sckobras** [comment]
Indeed current Samba needs additional plumbing initially, but I wouldn't call `msktutil`'s Samba support unusable. Try something like
<code>
net ads testjoin || net getdomainsid | grep -qw MY_DOMAIN || net -P -U MYHOSTNAME$ rpc getsid -S mydc.example.com
</code>
before calling `msktutil` with `--set-samba-secret` the first time, and you should be fine. I'm undecided whether we should integrate these calls into `msktutil` proper, though.

**Comment by jaroslawp** [comment]
Well - unusable without further instructions 

OK:  I can get and store Domain SID this way:
```
net -k -P rpc getsid -S mydc.example.com
Storing SID S-1-5-xxxxx[..]- for Domain MY_DOMAIN in secrets.tdb
```
(after a valid keytab has been obtained with previous msktutil run)
but this does not solve problem entirely:

```
# net changesecretpw -f
Enter machine password: 
secrets_prepare_password_change: secrets_fetch_or_upgrade_domain_info(DOMAIN) failed
Unable to write the machine account password in the secrets database
```
.. would you have an idea what this kind of problem could be ? 
(making the command more verbose / add debug option does not bring
more information)

Many thanks for help!

**Comment by sckobras** [comment]
The above sequence is the gist of what we're using in production with `msktutil` and Samba 4.6. We haven't tried with 4.7, yet, which might be even more picky. Could it be that you're hitting a variation of  [RH#1583252](https://bugzilla.redhat.com/show_bug.cgi?id=1583252)? I don't have access to its duplicate bug, but from what I've heard it might be due to uninitialized values in `secrets.tdb`. Just to confirm: does the issue go away if you copy over a secrets.tdb from an already joined machine, and fill all its entries with bogus values (something like
<code>
tdbtool secrets.tdb keys | awk '{print $4}' | xargs -n 1 -i tdbtool secrets.tdb store {} 'argh\0'
</code>
should do), and then run the above procedure including `net rpc getsid`?

**Comment by jaroslawp** [comment]
Indeed 4.7 seems to be more picky ! And you are right - the problem is with uninitialized values in secrets.tdb, here is what works for me (after getting/storing Domain SID):
```
tdbtool secrets.tdb store SECRETS/MACHINE_LAST_CHANGE_TIME/MY_DOMAIN 0000
tdbtool secrets.tdb store SECRETS/MACHINE_PASSWORD.PREV/MY_DOMAIN none
tdbtool secrets.tdb store SECRETS/MACHINE_PASSWORD/MY_DOMAIN none
```
(of course any string can be used in place of none)

As for RH#1583252 -> yes 4.7.1 is affected I'm using it with 4.7 patch from here: https://bugzilla.samba.org/show_bug.cgi?id=13376
which prevents segfault. So these are not official RH packages and possibly upcoming RH
patch will fix more. or maybe this is a samba bug and changesecretpw implementation shall be changed ? ...

Thanks !

**Comment by dkobras** [comment]
Cool, thanks for the additional info! From the `msktutil` side, I've just created PR #127, so at least we don't leave the system with a broken keytab and a broken Samba if `changesecretpw` fails. Not really a fix for this issue, but at least the system should be able in principle to recover without admin intervention.

**Comment by dkobras** [comment]
In order to solve this issue, I'd suggest to remove the Samba interaction from the C++ code completely, and replace it with a generic mechanism that calls one (or more?) executables and supplies them with the new password on `stdin`. We can add a simple
<code>
#!/bin/sh
net changesecretpw -f -i
</code>
as the default callout, thus retaining the current behavior of `--set-samba-secret`,  but admins could easily extend it as required, like the `tdbtool` workarounds given above, or an additional `net setdomainsid`. I'd rather not put all this magic into the `msktutil` binary itself.

@mproehl what do you think?

**Comment by mproehl** [comment]
@sckobras I like the idea of an external script instead of hard-coding "net ads ..." commands in msktutil

**Comment by michaelweiser** [comment]
@dkobras & @mproehl: Just spent an afternoon coming up with the same solution for the changesecretpw problem (again) to then find the solution ready to go here. ;)

What would this hooking mechanism look like exactly? I guess as a first step, --set-samba-secret and --use-samba-command-cmd could be merged into a --post-secret-update-cmd.

What further hooks do you envision?

**Comment by michaelweiser** [comment]
FWIW: ``net -k -P rpc getsid`` doesn't actually seem to do Kerberos, mumbles something about NTLM1 and fails in my environment. The following workaround did the trick:
``` shell
SID=$(rpcclient -k -c "lookupnames $WORKGROUP" $DCHOST)
[ -z "$SID" ] && echo "Error looking up domain SID" && exit 1

$NET setdomainsid $SID
```

**Comment by dkobras** [comment]
@michaelweiser Apart from keeping Samba in sync, an appropriate hook/callout mechanism could also provide an easy way to adjust further LDAP attributes. This is a recurring theme, and having to implement each request within the binary itself is a rather clumsy and inflexible way to address it. Also, one might want to keep a copy of the machine credentials in a central vault. And once you're there, it becomes viable to provide similar pre-hooks that can acquire credentials from further sources (Samba, password vault, etc.).

**Comment by michaelweiser** [comment]
PRs welcome, I assume?

One thing I ran into yesterday was a need to know the workgroup, selected DC and maybe other domain and connection parameters in the hook. I guess we should provide it/them with appropriate positional parameters or environment variables (e.g. for the password to keep it somewhat obscured)? I'd just go ahead and see what's available at the respective points in the program and pass it on.

**Comment by dkobras** [comment]
> PRs welcome, I assume?

@michaelweiser Yes and no, actually, because I didn't mention the downside yet, and that's the nagging voice that tells me 'second system effect'. So far, simplicity is one of msktutil's main advantages: get it compiled on an obscure arch, take the binary, and you're good to go. Your comment actually had me rethink: Maybe putting a hooks feature into msktutil proper may be putting the cart before the horse, and there's a simpler way to do things? Anyway, we're drifting away from the original subject of this issue, so I've opened #139 to summarize my ideas. Feel free to comment!

**Comment by slowfranklin** [comment]
For the records, this seems to be about https://bugzilla.samba.org/show_bug.cgi?id=13577

**Comment by max-tremblay** [comment]
For those who encounter this issue and are still not able to fix the problem by following previous solutions, I have found a workaround by setting the domain SID manually:

`net setdomainsid <DOMAIN_SID>`

We have then removed our machine account from AD, regenerated the keytab and executed msktutil
